packages/browser/src/core/page/get-page-context.ts

```diff
@@ -0,0 +1,141 @@
+import { isPlainObject } from '@segment/analytics-core'
+
+/**
+ * Final Page Context object expected in the Segment Event context
+ */
+export interface PageContext {
+  path: string
+  referrer: string
+  search: string
+  title: string
+  url: string
+}
+
+type CanonicalUrl = string | undefined
+
+export const BufferedPageContextDiscriminant = 'bpc' as const
+/**
+ * Page Context expected to be built by the snippet.
+ * Note: The key names are super short because we want to keep the strings in the html snippet short to save bytes.
+ */
+export interface BufferedPageContext {
+  __t: typeof BufferedPageContextDiscriminant // for extra uniqeness
+  c: CanonicalUrl
+  p: PageContext['path']
+  u: PageContext['url']
+  s: PageContext['search']
+  t: PageContext['title']
+  r: PageContext['referrer']
+}
+
+/**
+ * `BufferedPageContext` object builder
+ */
+export const createBufferedPageContext = (
+  url: string,
+  canonicalUrl: CanonicalUrl,
+  search: string,
+  path: string,
+  title: string,
+  referrer: string
+): BufferedPageContext => ({
+  __t: BufferedPageContextDiscriminant,
+  c: canonicalUrl,
+  p: path,
+  u: url,
+  s: search,
+  t: title,
+  r: referrer,
+})
+
+// my clever/dubious way of making sure this type guard does not get out sync with the type definition
+const BUFFERED_PAGE_CONTEXT_KEYS = Object.keys(
+  createBufferedPageContext('', '', '', '', '', '')
+) as (keyof BufferedPageContext)[]
+
+export function isBufferedPageContext(
+  bufferedPageCtx: unknown
+): bufferedPageCtx is BufferedPageContext {
+  if (!isPlainObject(bufferedPageCtx)) return false
+  if (bufferedPageCtx.__t !== BufferedPageContextDiscriminant) return false
+
+  for (const k of BUFFERED_PAGE_CONTEXT_KEYS) {
+    if (!(k in bufferedPageCtx)) return false
+  }
+  // if object has both the correct key names AND the right number of keys, assume it's a BufferedPageContext
+  if (
+    Object.keys(bufferedPageCtx).length !== BUFFERED_PAGE_CONTEXT_KEYS.length
+  ) {
+    return false
+  }
+  return true
+}
+
+//  Legacy logic: we are we appending search parameters to the canonical URL -- I guess the canonical URL is  "not canonical enough" (lol)
+const createCanonicalURL = (canonicalUrl: string, searchParams: string) => {
+  return canonicalUrl.indexOf('?') > -1
+    ? canonicalUrl
+    : canonicalUrl + searchParams
+}
+
+/**
+ * Strips hash from URL.
+ * http://www.segment.local#test -> http://www.segment.local
+ */
+const removeHash = (href: string) => {
+  const hashIdx = href.indexOf('#')
+  return hashIdx === -1 ? href : href.slice(0, hashIdx)
+}
+
+const formatCanonicalPath = (canonicalUrl: string) => {
+  const a = document.createElement('a')
+  a.href = canonicalUrl
+  // remove leading slash from canonical URL path
+  return a.pathname[0] === '/' ? a.pathname : '/' + a.pathname
```

Comment:
This is super legacy code -- I had the same question and did some testing, and a main difference is that new URL will throw errors (e.g. foo.com will throw an error because it doesn't have https))